staging/src/k8s.io/apimachinery/pkg/util/waitgroup/ratelimited_waitgroup_test.go

```diff
@@ -0,0 +1,227 @@
+	if count := target.test_Count(); count != n2 {
+		t.Errorf("expected count to be: %d, but got: %d", n2, count)
+	}
+
+	waitDoneCh := make(chan waitResult)
+	go func() {
+		factory.grace = 2 * time.Second
+		before, after, err := target.Wait(factory.NewRateLimiter)
+		waitDoneCh <- waitResult{before: before, after: after, err: err}
+	}()
+	// make sure there is no flake in the test due to this race condition
+	var waitModeGot bool
+	wait.PollImmediate(time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		if waiting := target.test_IsInWait(); waiting {
+			waitModeGot = true
+			return true, nil
+		}
+		return false, nil
+	})
+	// verify that wait mode has been triggered
+	if !waitModeGot {
+		t.Errorf("expected to be in waiting")
+	}
+
+	if err := target.Add(1); err == nil ||
+		!strings.Contains(err.Error(), "add with positive delta after Wait is forbidden") {
+		t.Errorf("expected Add to return error while in waiting mode: %v", err)
+	}
+	// make sure that RateLimitedSafeWaitGroup passes the right
+	// request count to the limiter factory.
+	if factory.countGot != n2 {
+		t.Errorf("expected count passed to factory to be: %d, but got: %d", n2, factory.countGot)
+	}
+
+	// unblock the second set to start calling Done
+	close(blockedCh)
+	wg2Done.Wait()
+	if invoked := limiter.invoked(); invoked != n2 {
+		t.Errorf("expected rate limiter to be called %d times, but got: %d", n2, invoked)
+	}
+
+	waitResult := <-waitDoneCh
+	if count := target.test_Count(); count != 0 {
+		t.Errorf("expected count to be zero, but got: %d", count)
+	}
+	if waitResult.before != n2 {
+		t.Errorf("expected count before Wait to be: %d, but got: %d", n2, waitResult.before)
+	}
+	if waitResult.after != 0 {
+		t.Errorf("expected count after Wait to be zero, but got: %d", waitResult.after)
+	}
+	if cancelInvoked != 1 {
+		t.Errorf("expected context cancel to be invoked once, but got: %d", cancelInvoked)
+	}
+}
+
+func TestRateLimitedSafeWaitGroupWithHardTimeeout(t *testing.T) {
+	target := RateLimitedSafeWaitGroup{}
+	n := 10
+	wg := sync.WaitGroup{}
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			target.Add(1)
+		}()
+	}
+
+	wg.Wait()
+	if count := target.test_Count(); count != n {
+		t.Errorf("expected count to be: %d, but got: %d", n, count)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	activeAt, activeNow, err := target.Wait(func(count int) (RateLimiter, context.Context, context.CancelFunc) {
+		return nil, ctx, cancel
+	})
+	if activeAt != n {
+		t.Errorf("expected active at Wait to be: %d, but got: %d", n, activeAt)
+	}
+	if activeNow != n {
+		t.Errorf("expected active after Wait to be: %d, but got: %d", n, activeNow)
+	}
+	if err != context.Canceled {
+		t.Errorf("expected error: %v, but got: %v", context.Canceled, err)
+	}
+}
+
+type waitResult struct {
+	before, after int
+	err           error
+}
+
+// used by test only
+func (wg *RateLimitedSafeWaitGroup) test_Count() int {
```

Comment:
Instead of this, can we go with sth like:
```
type TestRateLimitedSafeWaitGroup {
  RateLimitedSafeWaitGroup
}

func (wg *RateLimitedSafeWaitGroup) Count() int {
  ...
}

...
```

Reply:
that's a good idea, thanks!